### PR TITLE
Tiered tank and storage blocks, GUI improvements, refactoring, cell nerf and misc organisation

### LINKED ISSUE #1576: Digital Tank [Suggestion]
Just wondering if a digital tank could be made, Like there is the digital chest, quantum chest, there is a quantum tank, but no digital tank! Poor quantum tank is all alone, lol.

Idk how much it should hold. I guess not much? 

Thanks!

### LINKED ISSUE #1939: Description missing for interdimensional SU advancement
**Describe the bug**
Missing description for interdimensional SU advancement

**Screenshots**
image

**Environment (please complete the following information with the version):**
 - Minecraft: 1.15
 - Mod Loader: fabric

### PR DESCRIPTION
**Pull request by jusvit**
# Added
### Storage Units
- Crude Storage Unit
- Basic Storage Unit
- Advanced Storage Unit
- Industrial Storage Unit
- Quantum Storage Unit

### Tank Units
- Basic Tank Unit
- Advanced Tank Unit
- Industrial Tank Unit
- Quantum Tank Unit

### Storage Unit feature
- Right click a non-empty storage unit with the stack already inside of it, and it'll remove all stacks of that type from your inventory and deposit (stops if full)

### Misc crafting
- Basic display (New, using green variation of original texture, cheaper version)
- Digital display (Same)

### Textures
New textures for every storage & tank unit as well as a new basic display texture.
**Thanks to Spearkiller.**

### Advancements
- Some serious tankage (Place down a quantum tank unit)

# Changes
- Tidied up packages and moved stragglers into a more precise package
- Restricted cell max size to 16 (Nerfed so people might consider using new system)
- Various recipe changes around new unit system (**Needs feedback and tweaking**)
- Digital display has Basic display as part of recipe
- Rewrote majority of the (previously named) TechStorageBaseBlockEntity (now named) StorageUnitBaseBlockEntity
- Improved tooltips for storage and tank units

# Fixes
- Fix translation for interdimensionalSU advancement

## Footnote
Don't know if I missed anything that I've altered and didn't list above. I did test it all throughout developing, haven't seen any major flaws.

closes #1576 and closes #1939

### CODE REVIEW

**Comment by jusvit** [comment]
Currently moving it all to an ENUM, but please check over the storage unit base code

**Comment by jusvit** [comment]
Not ready, don't merge.

**Comment by jusvit** [comment]
Ready for review, configs and recipes might need to be tweaked

File: src/main/java/techreborn/blockentity/storage/fluid/CreativeQuantumTankBlockEntity.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+		Tank tank = this.getTank();
+
+		inventory.clear();
+
+		world.setBlockState(pos, TRContent.TankUnit.CREATIVE.block.getDefaultState());
+
+		TankUnitBaseBlockEntity tankEntity = (TankUnitBaseBlockEntity) world.getBlockEntity(pos);
+
+		tankEntity.setTank(tank);
```
Comment (drcrazy) [inline src/main/java/techreborn/blockentity/storage/fluid/CreativeQuantumTankBlockEntity.java:39]: Isn't this method available in creative tank also, thus you do not need to getBlockEntity additionally?

File: src/main/java/techreborn/blockentity/storage/fluid/TankUnitBaseBlockEntity.java [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+	@Override
+	public void addInfo(final List<Text> info, final boolean isReal, boolean hasData) {
+		if (isReal || hasData) {
+			if (!this.tank.getFluidInstance().isEmpty()) {
+				info.add(new LiteralText(this.tank.getFluidAmount() + " of " + this.tank.getFluid()));
+			} else {
+				info.add(new LiteralText("Empty"));
+			}
+		}
+		info.add(new LiteralText(Formatting.GRAY + "Capacity: " + Formatting.GOLD + this.tank.getCapacity() + " (" + this.tank.getCapacity().getRawValue() / 1000 + ")"));
```
Comment (drcrazy) [inline src/main/java/techreborn/blockentity/storage/fluid/TankUnitBaseBlockEntity.java]: PLease, add translation

File: src/main/java/techreborn/blockentity/storage/item/StorageUnitBaseBlockEntity.java [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+		this.maxCapacity = maxCapacity;
+		storeItemStack = ItemStack.EMPTY;
+		type = TRContent.StorageUnit.QUANTUM;
+		inventory = new RebornInventory<>(3, name, maxCapacity, this);
+	}
+
+	private void configureEntity(TRContent.StorageUnit type) {
+		this.maxCapacity = type.capacity;
+		storeItemStack = ItemStack.EMPTY;
+		inventory = new RebornInventory<>(2, "ItemInventory", 64, this);
```
Comment (drcrazy) [inline src/main/java/techreborn/blockentity/storage/item/StorageUnitBaseBlockEntity.java:92]: Is size 2 or 3?

File: src/main/java/techreborn/client/gui/GuiStorageUnit.java [hunk 4]
```diff
@@ -27,36 +27,48 @@
...
-		if (this.quantumChest.storedItem.isEmpty() && !this.quantumChest.inventory.getInvStack(1).isEmpty()) {
-			this.builder.drawBigBlueBar(this, 31, 43, this.quantumChest.inventory.getInvStack(1).getCount(), this.quantumChest.maxCapacity, mouseX - this.x, mouseY - this.y, "Stored", layer);
+		if (storageEntity.isEmpty()) {
+			font.draw("Empty", 10, 20, 4210752);
+		} else {
+			font.draw("Storing:", 10, 20, 4210752);
+			font.draw(storageEntity.getStoredStack().getName().asString(), 10, 30, 4210752);
+
+
+			font.draw("Amount:", 10, 50, 4210752);
```
Comment (drcrazy) [inline src/main/java/techreborn/client/gui/GuiStorageUnit.java]: Can you add translation, please?

File: src/main/java/techreborn/client/gui/GuiTankUnit.java [hunk 5]
```diff
@@ -29,40 +29,51 @@
...
-		FluidInstance fluid = quantumTank.tank.getFluidInstance();
-		if(fluid != null){
-			font.draw( "Fluid Type:", 10, 20, 4210752);
-			font.draw(FluidUtil.getFluidName(fluid) + "", 10, 30, 4210752);
+		FluidInstance fluid = tankEntity.getTank().getFluidInstance();
+
+		if (fluid.isEmpty()) {
+			font.draw("Empty", 10, 20, 4210752);
+		} else {
+			font.draw("Fluid Type:", 10, 20, 4210752);
```
Comment (drcrazy) [inline src/main/java/techreborn/client/gui/GuiTankUnit.java]: Can you add translation, please?

File: src/main/java/techreborn/items/tool/advanced/AdvancedJackhammerItem.java [hunk 6]
```diff
@@ -151,21 +151,21 @@ public boolean postMine(ItemStack stack, World worldIn, BlockState stateIn, Bloc
...
     public void usageTick(World world, LivingEntity entity,  ItemStack stack, int i) {
-        ItemUtils.checkActive(stack, cost, entity.world.isClient, MessageIDs.poweredToolID);
+      //  ItemUtils.checkActive(stack, cost, entity.world.isClient, MessageIDs.poweredToolID);
     }
 
     @Environment(EnvType.CLIENT)
     @Override
     public void appendTooltip(ItemStack stack, @Nullable World worldIn, List<Text> tooltip, TooltipContext flagIn) {
-        ItemUtils.buildActiveTooltip(stack, tooltip);
+       // ItemUtils.buildActiveTooltip(stack, tooltip);
```
Comment (drcrazy) [inline src/main/java/techreborn/items/tool/advanced/AdvancedJackhammerItem.java]: Should it be commented?

File: src/main/resources/data/techreborn/recipes/crafting_table/parts/basic_display.json [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+    "ADA",
+    "DCD",
+    "AGA"
+  ],
+  "key": {
+    "A": {
+      "item": "techreborn:refined_iron_plate"
+    },
+    "D": {
+      "item": "minecraft:ink_sac"
```
Comment (drcrazy) [inline src/main/resources/data/techreborn/recipes/crafting_table/parts/basic_display.json]: Can it be black dye?

File: src/main/resources/data/techreborn/recipes/crafting_table/parts/wood_plate.json [hunk 8]
```diff
@@ -0,0 +1,16 @@
+{
+  "type": "minecraft:crafting_shaped",
+  "pattern": [
+    "WWW",
+    "WWW",
+    "WWW"
+  ],
+  "key": {
+    "W": {
+      "tag": "minecraft:planks"
```
Comment (drcrazy) [inline src/main/resources/data/techreborn/recipes/crafting_table/parts/wood_plate.json]: Isn't it too much? I think we have compressor recipe for 1 plank -> 1 plate

**Review by drcrazy — COMMENTED** [review]
Looks good

File: src/main/java/techreborn/blockentity/storage/fluid/CreativeQuantumTankBlockEntity.java (hunk 1 above)
Comment (modmuss50) [reply src/main/java/techreborn/blockentity/storage/fluid/CreativeQuantumTankBlockEntity.java:39]: This seems fine, as `tankEntity` will be the new block entity at that location.

File: src/main/java/techreborn/blockentity/storage/fluid/TankUnitBaseBlockEntity.java (hunk 2 above)
Comment (modmuss50) [reply src/main/java/techreborn/blockentity/storage/fluid/TankUnitBaseBlockEntity.java]: Yes, few other places where its needed as well.

File: src/main/java/techreborn/blockentity/storage/item/QuantumChestBlockEntity.java [hunk 9]
```diff
@@ -22,40 +22,46 @@
...
+	public QuantumChestBlockEntity() {
+		super(TRBlockEntities.QUANTUM_CHEST, "QuantumChestBlockEntity", Integer.MAX_VALUE);
 	}
 
 	@Override
-	public IMachineGuiHandler getGui() {
-		return EGui.QUANTUM_CHEST;
+	public void onBreak(World world, PlayerEntity playerEntity, BlockPos blockPos, BlockState blockState) {
+		super.inventory.clear();
 	}
```
Comment (modmuss50) [inline src/main/java/techreborn/blockentity/storage/item/QuantumChestBlockEntity.java]: no call to super.onBreak

File: src/main/java/techreborn/blocks/storage/OldBlock.java [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+
+	public OldBlock(EGui gui, Supplier<BlockEntity> blockEntityClass, BlockState replacement) {
+		super(gui, blockEntityClass);
+		this.replacement = replacement;
+	}
+
+	@Override
+	public void onPlaced(World worldIn, BlockPos pos, BlockState state, LivingEntity placer, ItemStack stack) {
+		super.onPlaced(worldIn, pos, state, placer, stack);
+		worldIn.setBlockState(pos, replacement);
```
Comment (modmuss50) [inline src/main/java/techreborn/blocks/storage/OldBlock.java]: is this needed if you are doing the on tick? and no one would need to place the old block.

File: src/main/java/techreborn/client/gui/GuiStorageUnit.java [hunk 11]
```diff
@@ -27,36 +27,48 @@
...
 	}
 
 	@Override
 	protected void drawBackground(final float f, final int mouseX, final int mouseY) {
 		super.drawBackground(f, mouseX, mouseY);
 		final Layer layer = Layer.BACKGROUND;
 
-		drawSlot(80, 24, layer);
-		drawSlot(80, 64, layer);
+		drawString("IN", 100, 43, 4210752, layer);
```
Comment (modmuss50) [inline src/main/java/techreborn/client/gui/GuiStorageUnit.java]: Translation here as well.

File: src/main/java/techreborn/items/tool/advanced/AdvancedJackhammerItem.java (hunk 6 above)
Comment (modmuss50) [reply src/main/java/techreborn/items/tool/advanced/AdvancedJackhammerItem.java]: yeah, these changes seem wrong.

File: src/main/java/techreborn/blockentity/storage/item/StorageUnitBaseBlockEntity.java (hunk 3 above)
Comment (jusvit) [reply src/main/java/techreborn/blockentity/storage/item/StorageUnitBaseBlockEntity.java:92]: The way it was previously done was an input and output slot for the inventory (Some reason being 3 when the third is never used), and then a ItemStack variable. I just went with this method so it was easier get old data from saves.

File: src/main/java/techreborn/items/tool/advanced/AdvancedJackhammerItem.java (hunk 6 above)
Comment (jusvit) [reply src/main/java/techreborn/items/tool/advanced/AdvancedJackhammerItem.java]: Whoops

**Review by modmuss50 — APPROVED** [review]

**Review by drcrazy — APPROVED** [review]

**Pull request merged**